File: bigtangle-typescript/src/net/bigtangle/core/NetworkParameters.ts
import { Block } from './Block';
import { Sha256Hash } from './Sha256Hash';
 
import { ScriptBuilder } from '../script/ScriptBuilder';
import { Utils } from '../utils/Utils';
import { BitcoinSerializer } from './BitcoinSerializer'; // Import BitcoinSerializer directly
import { BigInteger } from './BigInteger';
import { ProtocolVersion } from './ProtocolVersion';
import { Transaction } from './Transaction';
import { TransactionInput } from './TransactionInput';
import { RewardInfo } from './RewardInfo';
import { ECKey } from './ECKey';
import { TransactionOutput } from './TransactionOutput';
import { Coin } from './Coin';
export abstract class NetworkParameters {
    static readonly ID_MAINNET = 'Mainnet';
    static readonly ID_UNITTESTNET = 'Test';
    protected    defaultSerializer: any;
    protected genesisBlock: Block | null = null;
    protected maxTarget: BigInteger = new BigInteger('0');
    protected maxTargetReward: BigInteger = new BigInteger('0');
    protected packetMagic: number = 0;
    protected addressHeader: number = 0;
    protected p2shHeader: number = 0;
    protected dumpedPrivateKeyHeader: number = 0;
    protected alertSigningKey: Uint8Array = new Uint8Array();
    protected bip32HeaderPub: number = 0;
    protected bip32HeaderPriv: number = 0;
    protected id: string = '';
    protected spendableCoinbaseDepth: number = 0;
    protected subsidyDecreaseBlockCount: number = 0;
    protected acceptableAddressCodes: number[] = [];
    protected dnsSeeds: string[] = [];
    protected addrSeeds: number[] = [];

    protected genesisPub: string = '';
    protected permissionDomainname: string[] = [];
  

    static readonly CONFIRMATION_UPPER_THRESHOLD_PERCENT = 51;
    static readonly CONFIRMATION_LOWER_THRESHOLD_PERCENT = 45;
    static readonly NUMBER_RATING_TIPS = 10;
    static readonly CONFIRMATION_UPPER_THRESHOLD = NetworkParameters.CONFIRMATION_UPPER_THRESHOLD_PERCENT * NetworkParameters.NUMBER_RATING_TIPS / 100;
    static readonly CONFIRMATION_LOWER_THRESHOLD = NetworkParameters.CONFIRMATION_LOWER_THRESHOLD_PERCENT * NetworkParameters.NUMBER_RATING_TIPS / 100;
    static readonly BIGTANGLE_TOKENID_STRING = 'bc';
    static readonly BIGTANGLE_TOKENNAME = 'BIG';
    static readonly BIGTANGLE_DECIMAL = 6;
    static readonly BLOCK_VERSION_GENESIS = 1;
    static readonly MAX_DEFAULT_BLOCK_SIZE = 1024 * 1024;
    static readonly MAX_BLOCK_SIGOPS = NetworkParameters.MAX_DEFAULT_BLOCK_SIZE / 50;
    static readonly BigtangleCoinTotal = "2100000000000000"; // Add this line, adjust value as needed
    static readonly ALLOWED_TIME_DRIFT = 5 * 60;
    static readonly HEADER_SIZE = 88 + 32 + 2 * 4 + 8 + 20 + 4 + 8;
    static readonly ORDER_TIMEOUT_MAX = 8 * 60 * 60;
    static readonly TARGET_TIMESPAN = 3 * 60 * 60;
    static readonly TARGET_SPACING = 30;
    static readonly INTERVAL = NetworkParameters.TARGET_TIMESPAN / NetworkParameters.TARGET_SPACING;
    static readonly TARGET_MAX_TPS = 100;
    static readonly TARGET_MAX_BLOCKS_IN_REWARD = 5000;
    static readonly MAX_REWARD_BLOCK_SIZE = NetworkParameters.MAX_DEFAULT_BLOCK_SIZE + NetworkParameters.TARGET_MAX_BLOCKS_IN_REWARD * 200;
    static readonly MILESTONE_CUTOFF = 40;
    static readonly FORWARD_BLOCK_HORIZON = NetworkParameters.TARGET_MAX_BLOCKS_IN_REWARD / 2 * 2;

    getId(): string { return this.id; }
    getSpendableCoinbaseDepth(): number { return this.spendableCoinbaseDepth; }
    getDnsSeeds(): string[] { return this.dnsSeeds; }
    getAddrSeeds(): number[] { return this.addrSeeds; }
    getGenesisBlock(): Block | null { return NetworkParameters.createGenesis(this); }
   
    getPacketMagic(): number { return this.packetMagic; }
    getAddressHeader(): number { return this.addressHeader; }
    getP2SHHeader(): number { return this.p2shHeader; }
    getDumpedPrivateKeyHeader(): number { return this.dumpedPrivateKeyHeader; }
    getAcceptableAddressCodes(): number[] { return this.acceptableAddressCodes; }
    getMaxTarget(): bigint { return this.maxTarget; }
    getMaxTargetReward(): bigint { return this.maxTargetReward; }
    getAlertSigningKey(): Uint8Array { return this.alertSigningKey; }
    getBip32HeaderPub(): number { return this.bip32HeaderPub; }
    getBip32HeaderPriv(): number { return this.bip32HeaderPriv; }
    getSubsidyDecreaseBlockCount(): number { return this.subsidyDecreaseBlockCount; }
 
    getPermissionDomainname(): string[] { return this.permissionDomainname; }
    getGenesisPub(): string { return this.genesisPub; }
    getDefaultSerializer(): any {
        this.defaultSerializer ??= new BitcoinSerializer(this, false);
        return this.defaultSerializer;
    }
    getOrderTimeoutMax(): number { return NetworkParameters.ORDER_TIMEOUT_MAX; }
    getTargetTimespan(): number { return NetworkParameters.TARGET_TIMESPAN; }
    getTargetSpacing(): number { return NetworkParameters.TARGET_SPACING; }
    getInterval(): number { return NetworkParameters.INTERVAL; }
    getTargetMaxTps(): number { return NetworkParameters.TARGET_MAX_TPS; }
    getTargetMaxBlocksInReward(): number { return NetworkParameters.TARGET_MAX_BLOCKS_IN_REWARD; }
    getMaxRewardBlockSize(): number { return NetworkParameters.MAX_REWARD_BLOCK_SIZE; }
    getMilestoneCutoff(): number { return NetworkParameters.MILESTONE_CUTOFF; }
    getForwardBlockHorizon(): number { return NetworkParameters.FORWARD_BLOCK_HORIZON; }

    static createGenesis(params: NetworkParameters): Block {
        const genesisBlock = new Block(
            params,
            (Sha256Hash as any).ZERO_HASH, // adapt as needed
            (Sha256Hash as any).ZERO_HASH,
            0, // Block.Type.BLOCKTYPE_INITIAL.ordinal() or equivalent
            0,
            0,
            Utils.encodeCompactBits(new BigInteger(params.getMaxTarget().toString()))
        );
        genesisBlock.setTime(1532896109);
        genesisBlock.setDifficultyTarget(Utils.encodeCompactBits(new BigInteger(params.getMaxTarget().toString())));
        const coinbase = new Transaction(params);
        const inputBuilder = new ScriptBuilder();
        coinbase.addInput(new TransactionInput(params, coinbase, Buffer.from(inputBuilder.build().getProgram())));
        const rewardInfo = new RewardInfo((Sha256Hash as any).ZERO_HASH, Utils.encodeCompactBits(new BigInteger(params.getMaxTargetReward().toString())), new Set(), 0);
        coinbase.setData(Buffer.from(rewardInfo.toByteArray()));
        // Add the coinbase output to the transaction
        const scriptPubKey = ScriptBuilder.createOutputScript(ECKey.fromPublic(params.genesisPub));
        coinbase.addOutput(new TransactionOutput(
            params,
            coinbase,
            new Coin(BigInt(NetworkParameters.BigtangleCoinTotal.toString()), Buffer.from(NetworkParameters.BIGTANGLE_TOKENID_STRING)),
            Buffer.from(scriptPubKey.getProgram())
        ));
        genesisBlock.addTransaction(coinbase);
        genesisBlock.setNonce(0);
        genesisBlock.setHeight(0);
        return genesisBlock;
    }

    // Static utility: fromID
    static fromID(id: string): NetworkParameters | null {
        if (id === NetworkParameters.ID_MAINNET) {
            return (require('../params/MainNetParams').MainNetParams).get();
        } else if (id === NetworkParameters.ID_UNITTESTNET) {
            return (require('../params/MainNetParams').MainNetParams).get();
        } else {
            return null;
        }
    }

    // Static utility: add
    static add(params: NetworkParameters, amount: bigint, account: string, coinbase: any): void {
        const list = account.split(',');
        const base = new Coin(amount, Buffer.from(NetworkParameters.BIGTANGLE_TOKENID_STRING));
        const keys: any[] = [];
        for (const s of list) {
            keys.push(ECKey.fromPublicOnly((Utils as any).HEX.decode(s.trim())));
        }
        if (keys.length <= 1) {
            coinbase.addOutput(new TransactionOutput(params, coinbase, base, Buffer.from(ScriptBuilder.createOutputScript(keys[0]).getProgram())));
        } else {
            const scriptPubKey = ScriptBuilder.createMultiSigOutputScript(keys.length - 1, keys);
            coinbase.addOutput(new TransactionOutput(params, coinbase, base, Buffer.from(scriptPubKey.getProgram())));
        }
    }

    // Instance methods (getters) already implemented above
    // equals and hashCode
    equals(o: any): boolean {
        if (this === o) return true;
        if (o === null || !(o instanceof NetworkParameters)) return false;
        return this.getId() === o.getId();
    }

    hashCode(): number {
        // Simple hash for string id
        let hash = 0, i, chr;
        const str = this.getId();
        if (str.length === 0) return hash;
        for (i = 0; i < str.length; i++) {
            chr = str.charCodeAt(i);
            hash = ((hash << 5) - hash) + chr;
            hash |= 0; // Convert to 32bit integer
        }
        return hash;
    }

    // getTransactionVerificationFlags (returns a Set or array of flags)
    getTransactionVerificationFlags(): Set<string> {
        // This is a simplification; adapt as needed for your Script.VerifyFlag
        return new Set(['P2SH', 'CHECKLOCKTIMEVERIFY']);
    }
getSerializer(parseRetain: boolean): BitcoinSerializer {
    return new BitcoinSerializer(this, parseRetain);
}
    getProtocolVersionNum( ): number {
        return ProtocolVersion.CURRENT; // Replace with the correct property, e.g., BITCOIN, if it exists
    }

    // Abstract methods
    abstract getUriScheme(): string;
    
   
    abstract serverSeeds(): string[];
    abstract getOrderPriceShift(orderBaseTokens: string): number;
    // Static utility methods (createGenesis, add, fromID, etc.) can be added as needed
}

File: bigtangle-core/src/main/java/net/bigtangle/core/NetworkParameters.java
/*******************************************************************************
 *  Copyright   2018  Inasset GmbH. 
 *  
 *******************************************************************************/
/*
 * Copyright 2011 Google Inc.
 * Copyright 2014 Andreas Schildbach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.bigtangle.core;

import static net.bigtangle.core.Utils.HEX;

import java.math.BigInteger;
import java.util.ArrayList;
import java.util.EnumSet;
import java.util.HashSet;
import java.util.List;

import javax.annotation.Nullable;

import com.google.common.base.Objects;
import com.google.common.math.LongMath;
 import net.bigtangle.params.MainNetParams;
import net.bigtangle.script.Script;
import net.bigtangle.script.ScriptBuilder;

/**
 * <p>
 * NetworkParameters contains the data needed for working with an instantiation
 * of a BigTangle.
 * </p>
 *
 * <p>
 * This is an abstract class, concrete instantiations can be found in the params
 * package. There are two: one for the main network ({@link MainNetParams}), one
 * for the public test network. Although this class contains some aliases for
 * them, you are encouraged to call the static get() methods on each specific
 * params class directly.
 * </p>
 */
public abstract class NetworkParameters {

    /**
     * The string returned by getId() for the main, production network where
     * people trade things.
     */
    public static final String ID_MAINNET = "Mainnet";

    /** Unit test network. */
    public static final String ID_UNITTESTNET = "Test";
 
    protected BigInteger maxTarget;
    protected BigInteger maxTargetReward;
    protected long packetMagic; // Indicates message origin network and is used
                                // to seek to the next message when stream state
                                // is unknown.
    protected int addressHeader;
    protected int p2shHeader;
    protected int dumpedPrivateKeyHeader;
    protected byte[] alertSigningKey;
    protected int bip32HeaderPub;
    protected int bip32HeaderPriv;

    /**
     * See getId(). This may be null for old deserialized wallets. In that case
     * we derive it heuristically by looking at the port number.
     */
    protected String id;

    /**
     * The depth of blocks required for a coinbase transaction to be spendable.
     */
    protected int spendableCoinbaseDepth;
    protected int subsidyDecreaseBlockCount;

    protected int[] acceptableAddressCodes;
    protected String[] dnsSeeds;
    protected int[] addrSeeds;

    protected transient MessageSerializer defaultSerializer = null;

    protected String genesisPub;
    //List of root permissionDomainname
    protected List<String> permissionDomainname;
    
    
    // MCMC settings
    public static final int CONFIRMATION_UPPER_THRESHOLD_PERCENT = 51;
    public static final int CONFIRMATION_LOWER_THRESHOLD_PERCENT = 45;
    public static final int NUMBER_RATING_TIPS = 10;
    public static final int CONFIRMATION_UPPER_THRESHOLD = CONFIRMATION_UPPER_THRESHOLD_PERCENT * NUMBER_RATING_TIPS
            / 100;
    public static final int CONFIRMATION_LOWER_THRESHOLD = CONFIRMATION_LOWER_THRESHOLD_PERCENT * NUMBER_RATING_TIPS
            / 100;

    // Token ID for System Coin
    public static final String BIGTANGLE_TOKENID_STRING = "bc";
    public static final byte[] BIGTANGLE_TOKENID = HEX.decode(BIGTANGLE_TOKENID_STRING);
    public static final String BIGTANGLE_TOKENNAME = "BIG";
    public static final int BIGTANGLE_DECIMAL = 6;
     
    protected int equihashN;
    protected int equihashK;
    /**
     * The version number at the start of the network.
     */
    public static final long BLOCK_VERSION_GENESIS = 1;

 
    /**
     * A constant shared by the entire network: how large in bytes a block is
     * allowed to be. 
     * It can no be smaller than last value, it will break consensus history. 
     * <p>
     * Start at: 262144 
     */
    public static final int MAX_DEFAULT_BLOCK_SIZE = 1024*1024; // 1MB
  
    /**
     * A "sigop" is a signature verification operation. Because they're
     * expensive we also impose a separate limit on the number in a block to
     * prevent somebody mining a huge block that has way more sigops than
     * normal, so is very expensive/slow to verify.
     */
    public static final int MAX_BLOCK_SIGOPS = MAX_DEFAULT_BLOCK_SIZE / 50;

    /**
     * The maximum allowed time drift of blocks into the future in seconds.
     */
    public static final long ALLOWED_TIME_DRIFT = 5 * 60;

    /**
     * How many bytes are required to represent a block header WITHOUT the
     * trailing 00 length byte.
     */
    public static final int HEADER_SIZE = 88 // bitcoin
            + 32 // additional branch prev block
            + 2 * 4 // time and difftarget from int to long
            + 8 // sequence (lastMiningReward) long
            + 20 // miner address
            + 4 // blockType
            + 8 // height
  ; 
    // max time of an order in seconds
    public static final long ORDER_TIMEOUT_MAX = 8 * 60 * 60;

    // 100 billions as Value
    public static BigInteger BigtangleCoinTotal = BigInteger.valueOf(LongMath.pow(10, 11 + BIGTANGLE_DECIMAL));
    public static final long TARGET_YEARLY_MINING_PAYOUT = BigtangleCoinTotal.longValue() / 1000;

    public static final int TARGET_TIMESPAN = 3 * 60 * 60;
    
    // cycle
    public static final int TARGET_SPACING = 30; // seconds per chain block.
    public static final int INTERVAL = TARGET_TIMESPAN / TARGET_SPACING;

    public static final int TARGET_MAX_TPS = 100;

    // per interval of length target_spacing, the reward is:
    public static final long TARGET_INTERVAL_REWARD = TARGET_YEARLY_MINING_PAYOUT * TARGET_SPACING / 31536000;
    // a third always comes directly from the consensus blocks:
    public static final long REWARD_AMOUNT_BLOCK_REWARD = TARGET_INTERVAL_REWARD / 3;
    // the other two thirds are for each inclusion into consensus and each block
    // itself:
    public static final long PER_BLOCK_REWARD = TARGET_INTERVAL_REWARD / 3 / TARGET_MAX_TPS / TARGET_SPACING;

    // block number can be taken in a reward block, it can not be too large for build reward.
    // about TARGET_MAX_TPS * TARGET_SPACING
    public static final int TARGET_MAX_BLOCKS_IN_REWARD = 5000;
    public static final int MAX_REWARD_BLOCK_SIZE = MAX_DEFAULT_BLOCK_SIZE + TARGET_MAX_BLOCKS_IN_REWARD * 200;

    // MCMC and reward milestones will only take the blocks back to this confirmed reward block
    public static final int MILESTONE_CUTOFF = 40;
    // Heuristic: MCMC will only continue up to this much above the current confirmed reward block
    private static final int ESTIMATED_TANGLE_DIAMETER = 2;
    public static final int FORWARD_BLOCK_HORIZON = TARGET_MAX_BLOCKS_IN_REWARD / ESTIMATED_TANGLE_DIAMETER * 2;

    
    protected NetworkParameters() {
    }

   

    /**
     * A Java package style string acting as unique ID for these parameters
     */
	public String getId() {
        return id;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o)
            return true;
        if (o == null || getClass() != o.getClass())
            return false;
        return getId().equals(((NetworkParameters) o).getId());
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(getId());
    }

    /**
     * Returns the network parameters for the given string ID or NULL if not
     * recognized.
     */
    @Nullable
    public static NetworkParameters fromID(String id) {
        if (id.equals(ID_MAINNET)) {
            return MainNetParams.get();
        } else if (id.equals(ID_UNITTESTNET)) {
            return MainNetParams.get();
        } else {
            return null;
        }
    }

    public int getSpendableCoinbaseDepth() {
        return spendableCoinbaseDepth;
    }

    /**
     * Returns DNS names that when resolved, give IP addresses of active peers.
     */
    public String[] getDnsSeeds() {
        return dnsSeeds;
    }

    /** Returns IP address of active peers. */
    public int[] getAddrSeeds() {
        return addrSeeds;
    }

 

    /** The header bytes that identify the start of a packet on this network. */
    public long getPacketMagic() {
        return packetMagic;
    }

    /**
     * First byte of a base58 encoded address. See
     * {@link net.bigtangle.core.Address}. This is the same as
     * acceptableAddressCodes[0] and is the one used for "normal" addresses.
     * Other types of address may be encountered with version codes found in the
     * acceptableAddressCodes array.
     */
    public int getAddressHeader() {
        return addressHeader;
    }

    /**
     * First byte of a base58 encoded P2SH address. P2SH addresses are defined
     * as part of BIP0013.
     */
    public int getP2SHHeader() {
        return p2shHeader;
    }

    /**
     * First byte of a base58 encoded dumped private key. See
     * {@link net.bigtangle.utils.DumpedPrivateKey}.
     */
    public int getDumpedPrivateKeyHeader() {
        return dumpedPrivateKeyHeader;
    }

    /**
     * The version codes that prefix addresses which are acceptable on this
     * network. Although Satoshi intended these to be used for "versioning", in
     * fact they are today used to discriminate what kind of data is contained
     * in the address and to prevent accidentally sending coins across chains
     * which would destroy them.
     */
    public int[] getAcceptableAddressCodes() {
        return acceptableAddressCodes;
    }

    /** Maximum target represents the easiest allowable proof of work. */
    public BigInteger getMaxTarget() {
        return maxTarget;
    }

    /** Maximum target represents the easiest allowable proof of work. */
    public BigInteger getMaxTargetReward() {
        return maxTargetReward;
    }

    /**
     * The key used to sign {@link net.bigtangle.core.AlertMessage}s. You can
     * use {@link net.bigtangle.core.ECKey#verify(byte[], byte[], byte[])} to
     * verify signatures using it.
     */
    public byte[] getAlertSigningKey() {
        return alertSigningKey;
    }

    /** Returns the 4 byte header for BIP32 (HD) wallet - public key part. */
    public int getBip32HeaderPub() {
        return bip32HeaderPub;
    }

    /** Returns the 4 byte header for BIP32 (HD) wallet - private key part. */
    public int getBip32HeaderPriv() {
        return bip32HeaderPriv;
    }

    // public abstract Coin getMaxMoney();

    /**
     * Scheme part for URIs, for example "bitcoin".
     */
    public abstract String getUriScheme();

    /**
     * Return the default serializer for this network. This is a shared
     * serializer.
     *
     */
    public final  synchronized MessageSerializer getDefaultSerializer() {
        // Construct a default serializer if we don't have one
        if (null == this.defaultSerializer) {
            // Don't grab a lock unless we absolutely need it
            synchronized (this) {
                // Now we have a lock, double check there's still no serializer
                // and create one if so.
                if (null == this.defaultSerializer) {
                    // As the serializers are intended to be immutable, creating
                    // two due to a race condition should not be a problem,
                    // however
                    // to be safe we ensure only one exists for each network.
                    this.defaultSerializer = getSerializer(false);
                }
            }
        }
        return defaultSerializer;
    }

    /**
     * Construct and return a custom serializer.
     */
    public abstract BitcoinSerializer getSerializer(boolean parseRetain);

    /**
     * The flags indicating which script validation tests should be applied to
     * the given transaction. Enables support for alternative blockchains which
     * enable tests based on different criteria.
     *
     */
    public EnumSet<Script.VerifyFlag> getTransactionVerificationFlags() {
        final EnumSet<Script.VerifyFlag> verifyFlags = EnumSet.noneOf(Script.VerifyFlag.class);
        // if (block.getTimeSeconds() >= NetworkParameters.BIP16_ENFORCE_TIME)
        verifyFlags.add(Script.VerifyFlag.P2SH);

        // Start enforcing CHECKLOCKTIMEVERIFY, (BIP65) for block.nVersion=4
        // blocks, when 75% of the network has upgraded:

        verifyFlags.add(Script.VerifyFlag.CHECKLOCKTIMEVERIFY);

        return verifyFlags;
    }

    public abstract int getProtocolVersionNum(final ProtocolVersion version);
    //initial server seeds to start, those server can register new servers and return other servers
    public abstract   String[]  serverSeeds() ;
    
    public enum ProtocolVersion {
        MINIMUM(70000), PONG(60001), BLOOM_FILTER(70000), CURRENT(70001);

        private final int bitcoinProtocol;

        ProtocolVersion(final int bitcoinProtocol) {
            this.bitcoinProtocol = bitcoinProtocol;
        }

        public int getBitcoinProtocolVersion() {
            return bitcoinProtocol;
        }
    }
    

    public List<PermissionDomainname> getPermissionDomainnameList() {
        ArrayList<PermissionDomainname> rootPermission = new ArrayList<>();
        for (String s : permissionDomainname) {
            rootPermission.add(new PermissionDomainname(s, ""));
        }
        return rootPermission;
    }

 
    /*
     * Order Price is in orderBaseToken and is used as Long, 
     * to enable the representation of value smaller than the unit of  orderBaseToken
     * this factor is used to shift the small price into long value.
     * It can not be changed after the initial set. 
     */
    public  abstract Integer getOrderPriceShift(String orderBaseTokens) ;

     
}
